torch/_subclasses/functional_tensor.py

```diff
@@ -158,6 +162,21 @@ def __new__(cls, elem):
         )
         torch._C._set_throw_on_mutable_data_ptr(out)
         out.elem = elem
+
+        if (
+            torch.is_inference_mode_enabled()
+            and torch._inductor.config.enable_auto_functionalized_v2
+        ):
+            if out.is_base_tensor():
```

Comment:
mentioned to @zou3519 offline, but I think it would be nice if we could kill this logic in favor of doing it directly in the C++ `FunctionalTensorWrapper`. in theory, we have all of the state that we need:

(1) current `FunctionalTensorWrapper` (view) tensor
(2) its `FunctionalStorageImpl::value`, aka the underlying (`FakeTensor`) base tensor

and performing the update for `updated_base, updated_view = auto_functionalize(mutable_op, base, view)` would look something like:

(1) sync the view to flush any pending mutations first
(2) perform `view_updated = mutable_op(clone(view)))`
(3) get the updated base with `updated_base = view_inverses(view_updated)`
(4) manually update the (shared) `FunctionalStorageImpl`, so its inner value gets replaced with `updated_base`, and update the counter on the storage so other aliases know the base was mutated

I think this should be mostly doable with the existing API's (e.g. `apply_updates`: https://github.com/pytorch/pytorch/blob/04118d86172e674865e9b6a12c1fbaa075fa7b00/aten/src/ATen/FunctionalStorageImpl.cpp#L132), although we might need to make a few tweaks

Reply:
@bdhirsh We tried some number of similar designs and ran into problems with all of them, but it's possible that we didn't find the right design. The problem with this particular approach is that `updated_base = view_inverses(view_updated)` emits a number of view_scatter operations into the graph that make it difficult to do the reinplacing later on (the reinplacing would need to know it can kill those view_scatter ops). The approach we've taken does not end up emitting view_scatter operations and it does this by doing
- Emit base_updated = auto_functionalized(mutable_op, base, view) into the graph (thereby hiding the scatters inside of auto_functionalized)
- replace/commit_update/sync on (base, base_updated)
- any views of base that get used in the future now are based on the `base_updated` Tensor. Using them will reconstruct the views, emitting view ops into the graph. These view ops don't need to be cleaned up later.